### ISSUE TITLE
Parameterisation of Skip/Take

### ISSUE THREAD

**Issue by stevendarby** [dotnet/efcore#29479]
Using this test as an example:
https://github.com/dotnet/efcore/blob/767bc93c6e6fca19868f486c7df8f4235bcf53cd/test/EFCore.Specification.Tests/Query/ComplexNavigationsCollectionsQueryTestBase.cs#L534-L553

There are two sets of Skip/Take in the query. This produces the following SQL:

https://github.com/dotnet/efcore/blob/d00fa6f55ad5be8986d365ef32f4cea905f59258/test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsCollectionsQuerySqlServerTest.cs#L1938-L1958

What is the rationale behind parameterising the first set but not the second? Or is this a bug? It was surprising to see a difference, as the way the Skip/Take are done in the LINQ looks the same. Thanks.

**Comment by roji** [comment]
@stevendarby the difference is a result of how expression trees work in C#.

When code is *inside* an expression tree lambda (e.g. the Select above), EF's LINQ provider gets a full expression tree representation of the C# code; this includes a different representation of constants/literals vs. parameters. EF uses SQL parameters (DbParameter) when the query expression tree contains a parameter, and SQL constants/literals when the expression tree contains a constant/literal.

However, when the LINQ operator is at the top-level - *not* inside an expression tree lambda (that's the case with the first Skip/Take operators above), EF doesn't get an expression tree, but only the actual value (1 above). That means we have no differentiation between parameter or constant, and so we can't generate SQL parameters or constants based on that.

If we always created SQL constants for this case, we'd generate different SQLs with different constants every time a different value is passed to top-level Skip/Take; that would be quite bad for perf. So the decision was made to always parameterize in that case, which is why you're seeing this translation difference.

Is there any particular reason you're asking about this, are you seeing some issue with the translation?

**Comment by stevendarby** [comment]
Thanks @roji, I think I follow that, mind if I repeat back my understanding to check it's right?

With the `.Skip(1).Take(3)` within the `Select`, you have the expression tree so you can respect what the developer has coded, which are constants in this case. If variables were used here, it would generate SQL parameters for them. In the first `.Skip(1).Take(1)` you just get the values. You have a choice between SQL parameters or constants and went with SQL parameters for perf reasons.

No issue with the translation - I came across this trying to tweak some of the parameterisation code for another issue and at one point had it (wrongly) parameterising some of the constants in the expression tree. This just drew my attention to the difference and thought there might be an issue in its own right. But understanding this properly now will help me with the other issue - thanks!

**Comment by roji** [comment]
@stevendarby yep, you got it.

For some extra bit of info... IIRC the old EF 6 (non-core) did translate top-level Skip/Take with SQL constants. To handle the perf issue with different SQLs (leading to different query plans), I think we had our own special overloads of Skip/Take which accept an expression lambda parameter, just so we could get an expression tree. So the user would be able to express constant vs. parameter by choosing different overloads; but in EF Core we made the decision to just always parameterize instead.

FYI we did receive one complaint that the Skip/Take parameterization is cause a perf degradation (#28632). If we confirm that, we *may* end up providing a mechanism for expressing constant parameters for top-level Skip/Take operators (but that's not planned at the moment).
